feat(storage): add support for Elasticsearch 8 (#2523)
kuzzleio/kuzzle: doc/2/guides/elasticsearch/v8/index.md

    content: Kuzzle, Documentation, kuzzle write pluggins, General purpose backend, iot, backend, opensource,  API Controllers
---

# Elasticsearch 8

<SinceBadge version="2.30.0"/>

Kuzzle uses Elasticsearch as a [NoSQL document store](/core/2/guides/main-concepts/data-storage).

With Kuzzle, customers **can directly access data stored in the database** as long as they have the rights to do so.

Kuzzle exposes the [Elasticsearch Query Language](/core/2/guides/main-concepts/querying) in a secure way. It is therefore possible to **take full advantage of the possibilities of Elasticsearch** with boolean queries, aggregations, special fields, etc.

The support of Elasticsearch 8 has been introduced in Kuzzle 2.30.0.

The choice has been made to keep kuzzle compatible to avoid breaking changes around the support of ES8.

We wanted to allow the user to OPT-IN for the feature so no modification is needed on your behalf to stay with Elasticsearch 7.

By default the majorVersion support will be 7 until Kuzzle v3.

The new key to change the version supported by is available under

```json
{
  "services": {
    "storageEngine": {
      "majorVersion": 8
    }
  }
}
```

:::warning

> You can not set the majorVersion to 8 if you are using a version of Kuzzle that does not support it.

Reviewer: ```suggestion
You can not set the `majorVersion` key to 8 if you are using a version of Kuzzle that does not support it.
```

:::

:::info

> Kuzzle cannot connect to both ES7 and ES8 at the same time.

Reviewer: ```suggestion
Kuzzle cannot connect to both Elasticsearch 7 and Elasticsearch 8 at the same time.
```

With Kuzzle, customers **can directly access data stored in the database** as long as they have the rights to do so.

Kuzzle exposes the [Elasticsearch Query Language](/core/2/guides/main-concepts/querying) in a secure way. It is therefore possible to **take full advantage of the possibilities of Elasticsearch** with boolean queries, aggregations, special fields, etc.

The support of Elasticsearch 8 has been introduced in Kuzzle 2.30.0.

The choice has been made to keep Kuzzle compatible to avoid breaking changes around the support of ES8. 

We wanted to allow the user to **opt-in** for this feature, so no modification is needed on your behalf if you want to keep using Elasticsearch 7.

The default major version of Elasticsearch will be 7 until Kuzzle v3.

The new configuration key to change the version supported is available under:  

```json
{
  "services": {
    "storageEngine": {
      "majorVersion": 8
    }
  }
}
```

:::warning
You can not set the `majorVersion` key to 8 if you are using a version of Kuzzle that does not support it. 
:::

:::info
Kuzzle cannot connect to both Elasticsearch 7 and Elasticsearch 8 at the same time.  

> :::

Reviewer 2: I guess it's not ::: but ``` here on these 7 lines

Author: This is part of the kuzzle documentation paradigm :)

* Elasticsearch Deprecation API can be used to check for any features or settings in your current cluster that are deprecated or removed in the 8.x version. Address these issues before proceeding.
* Test in a Non-Production Environment
Conduct a dry run in a development environment to spot potential issues and estimate the duration the migration process might take.

### Migration Methods

1. Re-indexing
	* Step 1: Create a new cluster running Elasticsearch 8.x.
	* Step 2: Take a snapshot of your data in the current 7.x cluster.
	* Step 3: Restore the snapshot into the new 8.x cluster.
1. Rolling Upgrade
	* Step 1: Disable Shard Allocation.
	* Step 2: Stop and upgrade a single Elasticsearch node.
	* Step 3: Enable Shard Allocation and allow the node to join the cluster and the cluster to re-balance.
	* Step 4: Repeat for each node in the cluster.
1. Post Upgrade Checks
	* Run the health and stats APIs to ensure the health of your newly upgraded cluster.
	* Update your clients and integrations to the latest version that's compatible with Elasticsearch 8.x, if not done already.
	* Monitor your cluster using the Monitoring API or third-party monitoring services.
1. Troubleshoot

>   * If you encounter any issues during the migration process, take advantage of the Elasticsearch documentation, forums, and issue trackers for troubleshooting information and support.

Reviewer 2: Align bullet point with the others

```suggestion
	* If you encounter any issues during the migration process, take advantage of the Elasticsearch documentation, forums, and issue trackers for troubleshooting information and support.
```
